File: packages/apple/Sources/BuiltIn/ClassifyChain.swift
import Foundation

/// Built-in chain for text classification.
///
/// ```swift
/// let result = try await ClassifyChain(categories: ["positive", "negative"]).run("Great product!")
/// print(result.topClassification.label) // "positive"
/// ```
@available(iOS 15.0, macOS 14.0, *)
public struct ClassifyChain: Chain {
    public let name = "ClassifyChain"

    private let model: any LocanaraModel
    private let categories: [String]
    private let maxResults: Int

    public init(
        model: (any LocanaraModel)? = nil,
        categories: [String] = ["positive", "negative", "neutral"],
        maxResults: Int = 3
    ) {
        self.model = model ?? LocanaraDefaults.model
        self.categories = categories
        self.maxResults = maxResults
    }

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        let prompt = try BuiltInPrompts.classify.format([
            "text": input.text,
            "categories": categories.joined(separator: ", ")
        ])

        let response = try await model.generate(prompt: prompt, config: .structured)
        let text = response.text.trimmingCharacters(in: .whitespacesAndNewlines)

        print("[ClassifyChain] raw response:\n---\n\(text)\n---")

        var classifications = text.components(separatedBy: "\n")
            .filter { !$0.isEmpty }
            .compactMap { line -> Classification? in
                let trimmed = line.trimmingCharacters(in: .whitespaces)
                guard let colonIndex = trimmed.lastIndex(of: ":") else { return nil }
                let label = String(trimmed[trimmed.startIndex..<colonIndex]).trimmingCharacters(in: .whitespaces).lowercased()
                let scoreStr = String(trimmed[trimmed.index(after: colonIndex)...]).trimmingCharacters(in: .whitespaces)
                let score = Double(scoreStr) ?? 0.0
                guard categories.contains(where: { $0.lowercased() == label }) else { return nil }
                return Classification(label: label, score: score)
            }

        // Fallback: if parsing found nothing, match against known categories
        if classifications.isEmpty {
            let lower = text.lowercased()
            if let match = categories.first(where: { lower.contains($0.lowercased()) }) {
                classifications = [Classification(label: match.lowercased(), score: 1.0)]
            } else {
                classifications = [Classification(label: text, score: 1.0)]
            }
        }

        print("[ClassifyChain] parsed: \(classifications.map { "\($0.label): \($0.score)" })")

        let sorted = Array(classifications.sorted { $0.score > $1.score }.prefix(maxResults))
        let result = ClassifyResult(
            classifications: sorted,
            topClassification: sorted[0]
        )

        return ChainOutput(
            value: result,
            text: sorted[0].label,
            metadata: input.metadata,
            processingTimeMs: response.processingTimeMs
        )
    }

    /// Type-safe execution that returns `ClassifyResult` directly.
    public func run(_ text: String) async throws -> ClassifyResult {
        let output = try await invoke(ChainInput(text: text))
        guard let result = output.typed(ClassifyResult.self) else {
            throw LocanaraError.executionFailed("Unexpected output type from ClassifyChain")
        }
        return result
    }
}

File: packages/android/locanara/src/main/kotlin/com/locanara/builtin/ClassifyChain.kt
package com.locanara.builtin

import com.locanara.Classification
import com.locanara.ClassifyResult
import com.locanara.composable.Chain
import com.locanara.core.ChainInput
import com.locanara.core.ChainOutput
import com.locanara.core.GenerationConfig
import com.locanara.core.LocanaraDefaults
import com.locanara.core.LocanaraModel

/**
 * Built-in chain for text classification.
 *
 * ```kotlin
 * val result = ClassifyChain(categories = listOf("positive", "negative")).run("Great product!")
 * println(result.topClassification.label) // "positive"
 * ```
 */
class ClassifyChain(
    private val model: LocanaraModel = LocanaraDefaults.model,
    private val categories: List<String> = listOf("positive", "negative", "neutral"),
    private val maxResults: Int = 3
) : Chain {
    override val name = "ClassifyChain"

    override suspend fun invoke(input: ChainInput): ChainOutput {
        val prompt = BuiltInPrompts.classify.format(
            mapOf("text" to input.text, "categories" to categories.joinToString(", "))
        )

        println("[ClassifyChain] input: ${input.text.take(200)}")
        val response = model.generate(prompt, GenerationConfig.STRUCTURED)
        val text = response.text.trim()
        println("[ClassifyChain] output: $text")

        var classifications = text.lines()
            .filter { it.isNotBlank() }
            .mapNotNull { line ->
                val trimmed = line.trim()
                val colonIndex = trimmed.lastIndexOf(':')
                if (colonIndex < 0) return@mapNotNull null
                val label = trimmed.substring(0, colonIndex).trim().lowercase()
                val scoreStr = trimmed.substring(colonIndex + 1).trim()
                val score = scoreStr.toDoubleOrNull() ?: return@mapNotNull null
                if (categories.none { it.lowercase() == label }) return@mapNotNull null
                Classification(label = label, score = score)
            }

        // Fallback: if parsing found nothing, match against known categories
        if (classifications.isEmpty()) {
            val lower = text.lowercase()
            val match = categories.firstOrNull { lower.contains(it.lowercase()) }
            classifications = listOf(Classification(label = (match ?: text).lowercase(), score = 1.0))
        }

        val sorted = classifications.sortedByDescending { it.score }.take(maxResults)
        val result = ClassifyResult(
            classifications = sorted,
            topClassification = sorted[0]
        )

        return ChainOutput(
            value = result,
            text = sorted[0].label,
            metadata = input.metadata,
            processingTimeMs = response.processingTimeMs
        )
    }

    /** Type-safe execution that returns [ClassifyResult] directly. */
    suspend fun run(text: String): ClassifyResult {
        val output = invoke(ChainInput(text = text))
        return output.typed<ClassifyResult>()
            ?: throw IllegalStateException("Unexpected output type from ClassifyChain")
    }
}
